src/documentation/execution-model.md

```diff
@@ -0,0 +1,163 @@
+---
+
+# Apache Beam Execution Model
+
+The Beam model allows runners to execute your pipeline in different ways.
+Depending on the runner’s choices, you may observe various effects as a result.
+This page describes the effects of these choices so you can better understand
+how Beam pipelines execute.
+
+**Table of Contents:**
+* TOC
+{:toc}
+
+## Processing of elements
+
+The serialization and communication of elements between machines is one of the
+most expensive operations in a distributed execution of your pipeline. Avoiding
+this serialization may require re-processing elements after failures or may
+limit the distribution of output to other machines.
+
+The runner processes elements on many machines and may serialize elements in
+between machines for other communication and persistence reasons.
+
+### Serialization and communication
+
+The runner may serialize elements for communication or persistence.
+
+The runner may decide to transfer elements between transforms in a variety of
+ways, such as:
+
+1.  Routing elements to a worker for processing as part of a grouping operation.
+    This may involve serializing elements and sorting them by their key.
+1.  Redistributing elements between workers to adjust parallelism. This may
+    involve serializing elements and communicating them to other workers.
+1.  Using the elements in a side input to a `ParDo`. This may require
+    serializing the elements and broadcasting them to all the workers executing
+    the `ParDo`.
+1.  Passing elements between transforms that are running on the same worker.
+    This may avoid having to serialize the elements, and instead just passing
+    them in memory.
+
+Additionally, the runner may serialize and persist elements for other reasons:
+
+1. When used as part of a Stateful `DoFn`, the runner may persist values to some
+   state mechanism.
+1. When committing the results of processing, the runner may persist the outputs
+   as a checkpoint.
+
+### Bundling and persistence
+
+Beam pipelines often focus on ["embarassingly parallel"](https://en.wikipedia.org/wiki/Embarrassingly_parallel)
+problems.  Because of this, the APIs emphasize processing elements in parallel,
+which makes it difficult to express things like "assign a sequence number to
+each element in a PCollection." This is intentional since such algorithms are
+much more likely to suffer from scalability problems.
+
+Processing all elements in parallel also has some drawbacks. Specifically, it
+makes it impossible to batch any operations, such as writing elements to a sink
+or checkpointing progress during processing.
+
+Instead of processing all elements simultaneously, the elements in a
+`PCollection` are processed in _bundles_. The division of the collection into
+bundles is arbitrary and selected by the runner. This allows the runner to
+choose an appropriate middle-ground between persisting results after every
+element, and having to retry everything if there is a failure.
```

Comment:
Would it help to have an example?

"For example, a streaming runner may process and commit small bundles while a batch runner may process larger bundles."